pkg/loki/loki.go

```diff
@@ -159,72 +170,145 @@ func (t *Loki) setupAuthMiddleware() {
+	if err != nil {
+		return err
 	}
-}
 
-func (t *Loki) stopModule(m moduleName) {
-	level.Info(util.Logger).Log("msg", "stopping", "module", m)
-	if modules[m].stop != nil {
-		if err := modules[m].stop(t); err != nil {
-			level.Error(util.Logger).Log("msg", "error stopping", "module", m, "err", err)
+	// before starting servers, register /ready handler. It should reflect entire Loki.
+	t.server.HTTP.Path("/ready").Handler(t.readyHandler(sm))
+
+	// Let's listen for events from this manager, and log them.
+	healthy := func() { level.Info(util.Logger).Log("msg", "Loki started") }
+	stopped := func() { level.Info(util.Logger).Log("msg", "Loki stopped") }
+	serviceFailed := func(service services.Service) {
+		// if any service fails, stop entire Loki
+		sm.StopAsync()
+
+		// let's find out which module failed
+		for m, s := range t.serviceMap {
+			if s == service {
+				if service.FailureCase() == util.ErrStopProcess {
+					level.Info(util.Logger).Log("msg", "received stop signal via return error", "module", m, "error", service.FailureCase())
+				} else {
+					level.Error(util.Logger).Log("msg", "module failed", "module", m, "error", service.FailureCase())
+				}
+				return
+			}
 		}
+
+		level.Error(util.Logger).Log("msg", "module failed", "module", "unknown", "error", service.FailureCase())
+	}
+
+	sm.AddListener(services.NewManagerListener(healthy, stopped, serviceFailed))
+
+	// Currently it's the Server that reacts on signal handler,
+	// so get Server service, and wait until it gets to Stopping state.
+	// It will also be stopped via service manager if any service fails (see attached service listener)
+	// Attach listener before starting services, or we may miss the notification.
+	serverStopping := make(chan struct{})
+	t.serviceMap[Server].AddListener(services.NewListener(nil, nil, func(from services.State) {
+		close(serverStopping)
+	}, nil, nil))
+
+	// Start all services. This can really only fail if some service is already
+	// in other state than New, which should not be the case.
+	err = sm.StartAsync(context.Background())
+	if err == nil {
+		// no error starting the services, now let's just wait until Server module
+		// transitions to Stopping (after SIGTERM or when some service fails),
+		// and then initiate shutdown
+		<-serverStopping
 	}
-}
 
-func (t *Loki) stopping(m moduleName) {
-	t.stoppingModule(m)
-	deps := orderedDeps(m)
-	// iterate over our deps in reverse order and call stoppingModule
-	for i := len(deps) - 1; i >= 0; i-- {
-		t.stoppingModule(deps[i])
+	// Stop all the services, and wait until they are all done.
+	// We don't care about this error, as it cannot really fail.
+	_ = services.StopManagerAndAwaitStopped(context.Background(), sm)
+
+	// if any service failed, report that as an error to caller
+	if err == nil {
+		if failed := sm.ServicesByState()[services.Failed]; len(failed) > 0 {
+			for _, f := range failed {
+				if f.FailureCase() != util.ErrStopProcess {
```

Comment:
is the error will be logged somewhere ? about which services has failed ?

Reply:
Earlier in the same method, there is `serviceFailed` listener which logs failure of any failed service.